_posts/2022-04-14-javascript-immutability.md

```diff
@@ -0,0 +1,452 @@
+---
+layout: post
+title: Writing Immutable JavaScript in 2022
+author: jarzka
+excerpt: >
+  JavaScript was not built for immutability, yet the concept has become more popular in recent years. How can we do it without cooking spaghetti?
+
+tags:
+ - JavaScript
+ - Immutability
+ - Immer
+ - ImmutableJS
+ - seamless-immutable
+ - lodash/fp
+ - Rambda
+---
+
+One of my colleagues once said that programming is just moving data from one place to another. From a database to a server, from the server to a client, from an application state to pixels on the screen. This data is rarely presented in one form: it needs to be modified, transformed, merged and split in order to be useful in different layers of an application.
+
+Once the data has been baked into a form that is ready to be used, you wouldn't like anyone to change it in the middle of the process, would you? Unfortunately, it is quite possible if your data is mutable. If you are using JavaScript, like many frontend developers, your data is mostly mutable by default - unprotected from an external change.
+
+I have been programming web frontend professionally for more than seven years now. What makes things interesting is that for most of my career I haven't touched JavaScript very much. I'm familiar with the language, but frontend apps that me and my teammates have built have been mainly written in [ClojureScript](https://clojurescript.org) - a Lisp dialect that compiles into JavaScript. Clojure has built-in support for functional programming and immutability, to which I am now highly addicted.
+
+I think this concept of immutability would be useful in the JavaScript world too. Unfortunately, JavaScript, by nature, is highly focused on object-oriented programming and mutable data. Luckily for us, it's still versatile enough for doing functional programming and modifying data immutably. Let's see how.
+
+## Brief Concept of Immutability
+
+Immutability is a simple concept: **an immutable value is some data that cannot be changed**. If you _do_ want to change it, you create a completely **new version of it with the updated data**.
```

Comment:
Sanoisin jotenkin näin niin ei jää epäilystä että alkuperäisellä ja modifioidulla olisi jokin maaginen yhteys_

you create **a completely new value**, which is equal to the original except for the modified parts.